Tests/Realm.Tests/Database/RealmValueWithCollections.cs

```diff
@@ -0,0 +1,1072 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using MongoDB.Bson;
+using NUnit.Framework;
+
+namespace Realms.Tests.Database
+{
+    [TestFixture, Preserve(AllMembers = true)]
+    internal class RealmValueWithCollections : RealmInstanceTest
+    {
+        public static Func<int, List<RealmValue>> ListGenerator = i => { return new List<RealmValue> { $"inner{i}", i }; };
+        public static Func<int, HashSet<RealmValue>> SetGenerator = i => { return new HashSet<RealmValue> { $"inner{i}", i }; };
+        public static Func<int, Dictionary<string, RealmValue>> DictGenerator = i => { return new Dictionary<string, RealmValue> { { "s1", i }, { "s2", $"ah{i}" } }; };
+
+        public static Func<int, RealmValue>[] CollectionGenerators = new Func<int, RealmValue>[]
+        {
+            i => (RealmValue)ListGenerator(i),
+            i => (RealmValue)SetGenerator(i),
+            i => (RealmValue)DictGenerator(i),
+        };
+
+        private RealmValueObject PersistAndFind(RealmValue rv)
+        {
+            _realm.Write(() =>
+            {
+                _realm.Add(new RealmValueObject { RealmValueProperty = rv });
+            });
+
+            return _realm.All<RealmValueObject>().First();
+        }
+
+        #region List
+
+        [Test]
+        public void List_WhenRetrieved_WorksWithAllTypes([Values(true, false)] bool isManaged)
+        {
+            var innerList = ListGenerator(1);
+            var innerSet = SetGenerator(1);
+            var innerDict = DictGenerator(1);
+
+            var originalList = new List<RealmValue>
+            {
+                RealmValue.Null,
+                1,
+                true,
+                "string",
+                new byte[] { 0, 1, 2 },
+                new DateTimeOffset(1234, 5, 6, 7, 8, 9, TimeSpan.Zero),
+                1f,
+                2d,
+                3m,
+                new ObjectId("5f63e882536de46d71877979"),
+                Guid.Parse("3809d6d9-7618-4b3d-8044-2aa35fd02f31"),
+                new InternalObject { IntProperty = 10, StringProperty = "brown" },
+                innerList,
+                innerSet,
+                innerDict,
+            };
+
+            RealmValue rv = originalList;
+
+            if (isManaged)
+            {
+                rv = PersistAndFind(rv).RealmValueProperty;
+            }
+
+            Assert.That(rv.Type, Is.EqualTo(RealmValueType.List));
+            Assert.That(rv != RealmValue.Null);
+
+            Assert.That(rv.AsList(), Is.EqualTo(originalList));
+            Assert.That(rv.AsAny(), Is.EqualTo(originalList));
+            Assert.That(rv.As<IList<RealmValue>>(), Is.EqualTo(originalList));
+
+            Assert.That(rv == originalList);
+            Assert.That(rv.Equals(originalList));
```

Comment:
This should probably also do an assert on `SequenceEquals` since as it is today, we can return `true` from `==/Equals` and this test would pass.

Reply:
We can be explicit about it, but this is already done by the assertions in the group before. `Is.Equal` underneath should be using something equivalent to `SequenceEquals`.

Reply:
I see, I think it's still better to be explicit about using `SequenceEquals` as `Is.EqualTo` typically uses the `.Equals` method (though I can see that for collections it's actually doing the right thing).

Reply:
Deal